### PR TITLE
fix: nanosim scikit-learn dependency fixed to pretrained-model version

### PR DESCRIPTION
**Pull request by dlaehnemann**
Pre-trained models of nanosim seem very tightly coupled to particular scikit-learn models, so this pinning syncs it to the currently stated version:
https://github.com/bcgsc/NanoSim/blob/bdcafb09a8fb8bfd6f3706423c8e69ef766e6a4a/README.md?plain=1#L95-L99

----

Please read the [guidelines for Bioconda recipes](https://bioconda.github.io/contributor/guidelines.html) before opening a pull request (PR).

### General instructions

* If this PR adds or updates a recipe, use "Add" or "Update" appropriately as the first word in its title.
* New recipes not directly relevant to the biological sciences need to be submitted to the [conda-forge channel](https://conda-forge.org/docs/) instead of Bioconda.
* PRs require reviews prior to being merged. Once your PR is passing tests and ready to be merged, please issue the `@BiocondaBot please add label` command.
* Please post questions [on Gitter](https://gitter.im/bioconda/Lobby) or ping `@bioconda/core` in a comment.

### Instructions for avoiding API, ABI, and CLI breakage issues
Conda is able to record and lock (a.k.a. pin) dependency versions used at build time of other recipes.
This way, one can avoid that expectations of a downstream recipe with regards to API, ABI, or CLI are violated by later changes in the recipe.
If not already present in the meta.yaml, make sure to specify `run_exports` (see [here](https://bioconda.github.io/contributor/linting.html#missing-run-exports) for the rationale and comprehensive explanation).
Add a `run_exports` section like this:

```yaml
build:
  run_exports:
    - ...

```

with `...` being one of:

| Case                             | run_exports statement                                               |
| -------------------------------- | ------------------------------------------------------------------- |
| semantic versioning              | `{{ pin_subpackage("myrecipe", max_pin="x") }}`     |
| semantic versioning (0.x.x)      | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}`   |
| known breakage in minor versions | `{{ pin_subpackage("myrecipe", max_pin="x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| known breakage in patch versions | `{{ pin_subpackage("myrecipe", max_pin="x.x.x") }}` (in such a case, please add a note that shortly mentions your evidence for that) |
| calendar versioning              | `{{ pin_subpackage("myrecipe", max_pin=None) }}`    |

while replacing `"myrecipe"` with either `name` if a `name|lower` variable is defined in your recipe or with the lowercase name of the package in quotes.

### Bot commands for PR management

<details>
  <summary>Please use the following BiocondaBot commands:</summary>

Everyone has access to the following BiocondaBot commands, which can be given in a comment:

<table>
  <tr>
    <td><code>@BiocondaBot please update</code></td>
    <td>Merge the master branch into a PR.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please add label</code></td>
    <td>Add the <code>please review & merge</code> label.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please fetch artifacts</code></td>
    <td>Post links to CI-built packages/containers. <br />You can use this to test packages locally.</td>
  </tr>
</table>

Note that the <code>@BiocondaBot please merge</code> command is now depreciated. Please just squash and merge instead.

Also, the bot watches for comments from non-members that include `@bioconda/<team>` and will automatically re-post them to notify the addressed `<team>`.

</details>

### CODE REVIEW

**Comment by dlaehnemann** [comment]
In addition, this PR updates the other dependencies based on the `nanosim` `requirements.txt`:
https://github.com/bcgsc/NanoSim/blob/bdcafb09a8fb8bfd6f3706423c8e69ef766e6a4a/requirements.txt

**Comment by dlaehnemann** [comment]
@lcoombe would you be willing to review this, as you probably know best if the dependency definitions make sense like this. One reason for this update is to make the version available on bioconda compatible with the latest pre-trained models you provided here:
https://github.com/bcgsc/NanoSim/pull/238

**Comment by lcoombe** [comment]
Thanks for this @dlaehnemann!
To make the recipe out-of-the-box compatible with the current pre-trained models, I agree that pinning scikit-learn makes sense. For the other requirements, I think that the existing constraints in the recipe (different from `requirements.txt` in the repo) are OK - I worry that they would be overly restrictive if we pin them all to specific versions. For example, there isn’t a need for minimap2 to be pinned to an older version, and there might be issues with installing NanoSim on different architectures as a result.

**Comment by dlaehnemann** [comment]
Thanks for having a look. The `~=` package matching means that newer patch versions of things will also be used, but major and minor version are pinned. So it is not just one version that is possible, but only a handful of versions that have the same or a higher patch version, see:
https://docs.conda.io/projects/conda-build/en/25.1.x/resources/package-spec.html#package-match-specifications

If you feel confident about newer versions for a number of tools, we could also move back to the `>=` specifications, that will only require a minimum version. However, this can become problematic with major version bumps that can contain breaking changes that might break the installation. So it might be necessary to either do the `~=x.x` matching only to the minor version, or also provide a (tested) upper limit with `=>x.x.x,<x`. If you have this from your tests, this would be great to use here.

And finally, if you want to make releasing NanoSim easier, including the bioconda recipe here, this guide looks like a well-developed walkthrough (including background explanations) for preparing a PyPI package, and PyPI packages are really easy to bundle and (almost) fully automatically update here on bioconda:
https://www.pyopensci.org/python-package-guide/index.html#python-packaging-for-scientists

Currently, I find the different version numbers in `requirements.txt` and different parts of the `README.md` a bit confusing. But I think instead of trying to harmonize that with a lot of manual effort, I would opt for the full packaging solution to make your life easier in the future.

**Comment by dlaehnemann** [comment]
OK, for now I'll keep it to the `scikit-learn` dependency fix over here, and leave everything else as is. We can then get this in sync with a consolidated dependency specification or even a PyPI package, should they come up at some point.

**Comment by lcoombe** [comment]
> Thanks for having a look. The `~=` package matching means that newer patch versions of things will also be used, but major and minor version are pinned. So it is not just one version that is possible, but only a handful of versions that have the same or a higher patch version, see: https://docs.conda.io/projects/conda-build/en/25.1.x/resources/package-spec.html#package-match-specifications
> 
> If you feel confident about newer versions for a number of tools, we could also move back to the `>=` specifications, that will only require a minimum version. However, this can become problematic with major version bumps that can contain breaking changes that might break the installation. So it might be necessary to either do the `~=x.x` matching only to the minor version, or also provide a (tested) upper limit with `=>x.x.x,<x`. If you have this from your tests, this would be great to use here.
> 
> And finally, if you want to make releasing NanoSim easier, including the bioconda recipe here, this guide looks like a well-developed walkthrough (including background explanations) for preparing a PyPI package, and PyPI packages are really easy to bundle and (almost) fully automatically update here on bioconda: https://www.pyopensci.org/python-package-guide/index.html#python-packaging-for-scientists
> 
> Currently, I find the different version numbers in `requirements.txt` and different parts of the `README.md` a bit confusing. But I think instead of trying to harmonize that with a lot of manual effort, I would opt for the full packaging solution to make your life easier in the future.

Thanks for the feedback! We don’t currently have a dedicated developer for NanoSim and I am on maternity leave, but I’ll ping @parham-k (a grad student in our group) so he’s aware of this back-and-forth.
But agreed for now, the changes look good and we’ll stick with the current maintenance of this recipe!

**Review by druvus — APPROVED** [review]

**Review by martin-g — APPROVED** [review]

**Comment by dlaehnemann** [comment]
Congrats and all the best for the maternity leave. Sorry for bothering you at this time, and many thanks for checking in, anyways. I always greatly appreciate the extra effort of keeping such useful software projects alive, especially without any dedicated developers!
